scylla-cql/src/serialize/value.rs

```diff
@@ -855,6 +879,66 @@ fn serialize_sequence<'t, 'b, T: SerializeValue + 't>(
         .map_err(|_| mk_ser_err_named(rust_name, typ, BuiltinSerializationErrorKind::SizeOverflow))
 }
 
+fn serialize_vector<'t, 'b, T: SerializeValue + 't>(
+    rust_name: &'static str,
+    len: usize,
+    iter: impl Iterator<Item = &'t T>,
+    element_type: &ColumnType,
+    dimensions: &u16,
+    typ: &ColumnType,
+    writer: CellWriter<'b>,
+) -> Result<WrittenCellProof<'b>, SerializationError> {
+    if len != *dimensions as usize {
+        return Err(mk_ser_err_named(
+            rust_name,
+            typ,
+            VectorSerializationErrorKind::InvalidNumberOfElements(len, *dimensions),
+        ));
+    }
+    let mut builder = writer.into_value_builder();
+    match element_type.type_size() {
+        Some(_) => {
+            for element in iter {
+                T::serialize(
+                    element,
+                    element_type,
+                    builder.make_sub_writer_without_size(),
+                )
+                .map_err(|err| {
+                    mk_ser_err_named(
+                        rust_name,
+                        typ,
+                        VectorSerializationErrorKind::ElementSerializationFailed(err),
+                    )
+                })?;
+            }
+        }
+        None => {
+            for element in iter {
+                let mut element_buffer = Vec::new();
+                let inner_writer = CellWriter::new_without_size(&mut element_buffer);
+                T::serialize(element, element_type, inner_writer).map_err(|err| {
+                    mk_ser_err_named(
+                        rust_name,
+                        typ,
+                        VectorSerializationErrorKind::ElementSerializationFailed(err),
+                    )
+                })?;
+                let mut element_length_buffer = Vec::new();
+                unsigned_vint_encode(
+                    element_buffer.len().try_into().unwrap(),
+                    &mut element_length_buffer,
+                );
+                builder.append_bytes(element_length_buffer.as_slice());
+                builder.append_bytes(element_buffer.as_slice());
+            }
```

Comment:
:recycle: Similarly to deserialization, let's extract separate functions for the constant length and variable length element cases.